### PR TITLE
docs: add docs and example showing how to get the expression data type

### LINKED ISSUE #7725: [Docs] How to get data type of a logical expression
### Is your feature request related to a problem or challenge?

It is not always apparent to new users how to get the data type for a logical expression.

### Describe the solution you'd like

Add some docs explaining ExprSchemable.

Quick example:

```rust
use arrow_schema::DataType;
use datafusion::common::DFField;
use datafusion::common::DFSchema;
use datafusion::logical_expr::col;
use datafusion::logical_expr::ExprSchemable;
use std::collections::HashMap;

fn main() {
    let expr = col("foo");
    let schema = DFSchema::new_with_metadata(
        vec![DFField::new_unqualified("foo", DataType::Utf8, true)],
        HashMap::new(),
    )
    .unwrap();
    print!("type = {}", expr.get_type(&schema).unwrap());
}
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by r3stl355**
## Which issue does this PR close?

Closes #7725

## Rationale for this change

Make it easier to understand what the `get_type` returns

## Are these changes tested?

An extra example in  `datafusion-examples/examples/expr_api.rs`
Docscring in `datafusion/expr/src/expr_schema.rs`
A paragraph in `docs/source/library-user-guide/working-with-exprs.md`

## Are there any user-facing changes?

Only documentation and an example, no breaking changes

### CODE REVIEW

**Comment by r3stl355** [comment]
Few issues with the docstrings. I can resolve some by using lib names instead of fully qualified (e.g. `datafusion_expr` instead of `datafusion::logical_expr`) but `arrow_schema::DataType` requires an `arrow-schema` dependency which I don't want to add to Cargo.toml just for this. Should I mark the docstring as `ignore`?

File: datafusion-examples/examples/expr_api.rs [hunk 1]
```diff
@@ -68,6 +71,9 @@ async fn main() -> Result<()> {
     // See how to analyze ranges in expressions
     range_analysis_demo()?;
 
+    // See how to get the type of the expression
```
Comment (alamb) [inline datafusion-examples/examples/expr_api.rs:74]:
```suggestion
    // See how to determine the data types of expressions
```

File: datafusion-examples/examples/expr_api.rs [hunk 2]
```diff
@@ -256,3 +262,38 @@ pub fn physical_expr(schema: &Schema, expr: Expr) -> Result<Arc<dyn PhysicalExpr
 
     create_physical_expr(&expr, df_schema.as_ref(), &props)
 }
+
+fn expression_type_demo() -> Result<()> {
```
Comment (alamb) [inline datafusion-examples/examples/expr_api.rs:266]:
```suggestion
/// This function shows how to use `Expr::get_type` to retrieve the DataType
/// of an expression. 
fn expression_type_demo() -> Result<()> {
```

File: datafusion-examples/examples/expr_api.rs [hunk 3]
```diff
@@ -256,3 +262,38 @@ pub fn physical_expr(schema: &Schema, expr: Expr) -> Result<Arc<dyn PhysicalExpr
...
+    let expr = col("c");
+
+    // Using a schema where the column `foo` is of type Utf8
```
Comment (alamb) [inline datafusion-examples/examples/expr_api.rs:269]:
```suggestion
    // To determine the DataType of an expression, DataFusion must know the 
    // types of the input expressions. You can provide this information using 
    // a schema. In this case we create a schema where the column `foo` is of 
    // type Utf8 (a String / VARCHAR)
```

File: datafusion-examples/examples/expr_api.rs [hunk 4]
```diff
@@ -256,3 +262,38 @@ pub fn physical_expr(schema: &Schema, expr: Expr) -> Result<Arc<dyn PhysicalExpr
...
+
+    // Using a schema where the column `foo` is of type Int32
+    let schema = DFSchema::new_with_metadata(
+        vec![DFField::new_unqualified("c", DataType::Int32, true)],
+        HashMap::new(),
+    )
+    .unwrap();
+    assert_eq!("Int32", format!("{}", expr.get_type(&schema).unwrap()));
+
+    // Get the type of an expression that adds 2 columns. Adding an Int32
```
Comment (alamb) [inline datafusion-examples/examples/expr_api.rs:285]: 👍

File: datafusion/expr/src/expr_schema.rs [hunk 5]
```diff
@@ -58,6 +58,60 @@ impl ExprSchemable for Expr {
...
+    ///   let schema = DFSchema::new_with_metadata(
+    ///     vec![DFField::new_unqualified("c", DataType::Int32, true)],
+    ///     HashMap::new(),
+    ///   )
+    ///   .unwrap();
+    ///   assert_eq!("Int32", format!("{}", expr.get_type(&schema).unwrap()));
+    /// }
+    /// ```
+    ///
+    /// ## Get the type of an expression that adds 2 columns. Adding an Int32
```
Comment (alamb) [inline datafusion/expr/src/expr_schema.rs:92]: I personally suggest we only keep one of the examples in the API docs (as I think the key point of the API is to show using a DF schema as `ExprSchemable`) 

Perhaps we only keep the example of adding two numbers together?

File: docs/source/library-user-guide/working-with-exprs.md [hunk 6]
```diff
@@ -180,6 +180,34 @@ Projection: Int64(1) + Int64(1) AS added_one
 
 I.e. the `add_one` UDF has been inlined into the projection.
 
+## Getting the data type of the expression
+
+The `arrow::datatypes::DataType` of the expression can be obtained by calling the `get_type` given the schema
```
Comment (alamb) [inline docs/source/library-user-guide/working-with-exprs.md:185]:
```suggestion
The `arrow::datatypes::DataType` of the expression can be obtained by calling the `get_type` given something that implements `Expr::Schemable`, for example a `DFschema` object:
```

**Review by alamb — APPROVED** [review]
Thank you @r3stl355  -- this looks great ❤️ . It will be very helpful

I had some small suggestions but I also think we could do them as follow on PRs (or never)

**Comment by r3stl355** [comment]
> I had some small suggestions but I also think we could do them as follow on PRs (or never)


@alamb  I'll make these changes later today, this is not merged yet so I don't see a reason for not implementing them

**Comment by alamb** [comment]
Thanks again @r3stl355

**Pull request merged**